applications/DEMApplication/custom_utilities/tesselation_utilities.cpp

```diff
@@ -0,0 +1,805 @@
+  {
+    KRATOS_TRY
+
+    ProcessInfo& r_process_info = r_modelpart.GetProcessInfo();
+
+    if (r_process_info[DOMAIN_SIZE] == 2)
+      Triangulation(r_modelpart);
+    else if (r_process_info[DOMAIN_SIZE] == 3)
+      Tetrahedralization(r_modelpart);
+    else
+      KRATOS_ERROR << "Invalid domain size!";
+
+    KRATOS_CATCH("")
+  }
+
+  //-----------------------------------------------------------------------------------------------------------------------
+  void TesselationUtilities::Triangulation(ModelPart& r_modelpart)
+  {
+    KRATOS_TRY
+
+    int num_particles = r_modelpart.NumberOfElements();
+    if (num_particles < 3)
+      return;
+
+    // Create and clear IO
+    struct triangulateio in, out, vorout;
+    ClearTriangle(in);
+    ClearTriangle(out);
+    ClearTriangle(vorout);
+
+    // Build input
+    in.numberofpoints = num_particles;
+    in.pointlist = (double*)malloc(sizeof(double) * in.numberofpoints * 2);
+
+    #pragma omp parallel for schedule(dynamic, 100)
+    for (int i = 0; i < in.numberofpoints; i++) {
+      ModelPart::ElementsContainerType::iterator it = r_modelpart.GetCommunicator().LocalMesh().Elements().ptr_begin() + i;
+      ThermalSphericParticle<SphericParticle>& particle = dynamic_cast<ThermalSphericParticle<SphericParticle>&> (*it);
+
+      particle.mDelaunayPointListIndex = i;
+
+      const array_1d<double, 3>& coors = particle.GetParticleCoordinates();
+      in.pointlist[2 * i + 0] = coors[0];
+      in.pointlist[2 * i + 1] = coors[1];
+
+      if (mUpdateVoronoi)
+        particle.mNeighborVoronoiRadius.clear();
+    }
+
+    // Perform triangulation
+    int fail = 0;
+    try {
+      triangulate(&mSwitches[0], &in, &out, &vorout);
+    }
+    catch (int error_code) {
+      fail = error_code;
+    }
+
+    if (fail || out.numberoftriangles == 0 || in.numberofpoints != out.numberofpoints) {
+      KRATOS_ERROR_IF(r_modelpart.GetProcessInfo()[TIME_STEPS] == 1) << "Fail to generate triangulation!" << std::endl;
+      KRATOS_WARNING("DEM") << std::endl;
+      KRATOS_WARNING("DEM") << "Fail to generate triangulation! Results from previous successful triangulation will be used." << std::endl;
+      KRATOS_WARNING("DEM") << std::endl;
+      FreeTriangle(in);
+      FreeTriangle(out);
+      FreeTriangle(vorout);
+      return;
+    }
+
+    // Update voronoi diagram
+    if (mUpdateVoronoi)
+      UpdateVoronoi2D(r_modelpart, out, vorout);
+
+    // Update average porosity
+    if (mUpdatePorosiy)
+      UpdatePorosity2D(r_modelpart, out, vorout);
+
+    // Free memory
+    FreeTriangle(in);
+    FreeTriangle(out);
+    FreeTriangle(vorout);
+
+    ClearTriangle(in);
+    ClearTriangle(out);
+    ClearTriangle(vorout);
+
+    KRATOS_CATCH("")
+  }
+
+  //-----------------------------------------------------------------------------------------------------------------------
+  void TesselationUtilities::Tetrahedralization(ModelPart& r_modelpart)
```

Comment:
If the variable is a function entry, should be `rModelPart`

Reply:
Name changed in every function of the class